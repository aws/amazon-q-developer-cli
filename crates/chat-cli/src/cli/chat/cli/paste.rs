use clap::Args;
use crossterm::execute;
use crossterm::style::{
    self,
    Color,
};

use crate::cli::chat::util::clipboard::paste_image_from_clipboard;
use crate::cli::chat::{
    ChatError,
    ChatSession,
    ChatState,
};
use crate::os::Os;

#[derive(Debug, Args, PartialEq)]
pub struct PasteArgs;

impl PasteArgs {
    pub async fn execute(self, _os: &mut Os, session: &mut ChatSession) -> Result<ChatState, ChatError> {
        match paste_image_from_clipboard() {
            Ok(path) => Ok(ChatState::HandleInput {
                input: path.display().to_string(),
            }),
            Err(e) => {
                execute!(
                    session.stderr,
                    style::SetForegroundColor(Color::Red),
                    style::Print("‚ùå Failed to paste image: "),
                    style::SetForegroundColor(Color::Reset),
                    style::Print(format!("{}\n", e))
                )?;

                Ok(ChatState::PromptUser {
                    skip_printing_tools: false,
                })
            },
        }
    }
}
