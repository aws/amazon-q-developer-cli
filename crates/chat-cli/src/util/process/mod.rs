#[cfg(target_os = "windows")]
mod windows;
#[cfg(target_os = "windows")]
pub use windows::*;

#[cfg(not(windows))]
mod unix;
