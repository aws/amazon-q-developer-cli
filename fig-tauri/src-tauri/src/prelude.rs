pub use anyhow::{anyhow, Result};
