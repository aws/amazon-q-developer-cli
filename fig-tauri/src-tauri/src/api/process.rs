use tokio::process::Command;

use fig_proto::fig::server_originated_message::Submessage as ServerOriginatedSubMessage;
use fig_proto::fig::{
    PseudoterminalExecuteRequest, PseudoterminalExecuteResponse, PseudoterminalWriteRequest,
    RunProcessRequest, RunProcessResponse,
};

use super::{ResponseKind, ResponseResult};
use crate::response_error;

// TODO(mia): implement actual pseudoterminal stuff
pub async fn execute(request: PseudoterminalExecuteRequest, _: i64) -> ResponseResult {
    let mut cmd = Command::new("/bin/bash");
    cmd.arg("--noprofile")
        .arg("--norc")
        .arg("-c")
        .arg(request.command)
        .current_dir(request.working_directory.unwrap_or_else(|| {
            std::env::current_dir()
                .expect("Failed getting current directory")
                .to_string_lossy()
                .to_owned()
                .to_string()
        }));
    for var in request.env {
        cmd.env(var.key.clone(), var.value());
    }
    let output = cmd
        .output()
        .await
        .map_err(response_error!("Failed running command"))?;

    Ok(ResponseKind::Message(Box::new(
        ServerOriginatedSubMessage::PseudoterminalExecuteResponse(PseudoterminalExecuteResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: if output.stderr.len() > 0 {
                Some(String::from_utf8_lossy(&output.stderr).to_string())
            } else {
                None
            },
            exit_code: output.status.code(),
        }),
    )))
}

pub async fn run(request: RunProcessRequest, _: i64) -> ResponseResult {
    let mut cmd = Command::new(request.executable);
    cmd.current_dir(request.working_directory.unwrap_or_else(|| {
        std::env::current_dir()
            .expect("Failed getting current directory")
            .to_string_lossy()
            .to_owned()
            .to_string()
    }));
    for arg in request.arguments {
        cmd.arg(arg);
    }
    for var in request.env {
        cmd.env(var.key.clone(), var.value());
    }

    let output = cmd
        .output()
        .await
        .map_err(response_error!("Failed running command"))?;

    Ok(ResponseKind::Message(Box::new(
        ServerOriginatedSubMessage::RunProcessResponse(RunProcessResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: String::from_utf8_lossy(&output.stderr).to_string(),
            exit_code: output.status.code().unwrap_or(0),
        }),
    )))
}

pub async fn write(_: PseudoterminalWriteRequest, _: i64) -> ResponseResult {
    Err(ResponseKind::Error(
        "PseudoterminalWriteRequest is deprecated".to_string(),
    ))
}
