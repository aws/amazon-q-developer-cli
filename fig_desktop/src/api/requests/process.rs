use std::process::Stdio;

use anyhow::anyhow;
use fig_proto::fig::server_originated_message::Submessage as ServerOriginatedSubMessage;
use fig_proto::fig::{
    EnvironmentVariable, PseudoterminalExecuteRequest, PseudoterminalExecuteResponse, RunProcessRequest,
    RunProcessResponse,
};
use tokio::process::Command;
use tracing::{debug, error, warn};

use super::{RequestResult, RequestResultImpl};
use crate::figterm::FigtermState;
use crate::native::SHELL;

fn get_shell_path_from_state(state: &FigtermState) -> Option<String> {
    state.most_recent_session()?.context?.shell_path
}

fn shell_args(shell_path: &str) -> &'static [&'static str] {
    let (_, shell_name) = shell_path
        .rsplit_once(|c| c == '/' || c == '\\')
        .unwrap_or(("", shell_path));
    match shell_name {
        "bash" | "bash.exe" => &["--norc", "--noprofile", "-c"],
        "zsh" | "zsh.exe" => &["--norcs", "-c"],
        "fish" | "fish.exe" => &["--no-config", "-c"],
        _ => {
            warn!("unknown shell {shell_name}");
            &[]
        },
    }
}

// TODO(mia): implement actual pseudoterminal stuff
pub async fn execute(request: PseudoterminalExecuteRequest, state: &FigtermState) -> RequestResult {
    let shell = get_shell_path_from_state(state).unwrap_or_else(|| SHELL.into());
    let args = shell_args(&shell);

    debug!(
        "Executing {:?} (shell {shell:?}, args {args:?}, cwd {:?}, env {:?})",
        request.command,
        request.working_directory(),
        request.env
    );

    let mut cmd = Command::new(shell);
    #[cfg(target_os = "windows")]
    cmd.creation_flags(windows::Win32::System::Threading::DETACHED_PROCESS.0);
    // TODO(sean): better SHELL_ARGs handling here based on shell.
    // TODO(sean): handle wsl distro from FigtermState here.
    cmd.args(args);
    cmd.stdin(Stdio::inherit());
    cmd.arg(&request.command);

    if let Some(working_directory) = request.working_directory {
        cmd.current_dir(working_directory);
    }

    cmd.env("FIG_ENV_VAR", "1");
    cmd.env("FIG_SHELL_VAR", "1");
    cmd.env("FIG_TERM", "1");
    cmd.env("FIG_PTY", "1");
    cmd.env("PROCESS_LAUNCHED_BY_FIG", "1");
    cmd.env("HISTFILE", "");
    cmd.env("HISTCONTROL", "ignoreboth");
    cmd.env("TERM", "xterm-256color");

    for EnvironmentVariable { key, value } in &request.env {
        match value {
            Some(value) => cmd.env(key, value),
            None => cmd.env_remove(key),
        };
    }

    let output = cmd
        .output()
        .await
        .map_err(|_| anyhow!("Failed running command: {:?}", request.command))?;

    RequestResult::Ok(Box::new(ServerOriginatedSubMessage::PseudoterminalExecuteResponse(
        PseudoterminalExecuteResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: if output.stderr.is_empty() {
                None
            } else {
                Some(String::from_utf8_lossy(&output.stderr).to_string())
            },
            exit_code: output.status.code(),
        },
    )))
}

pub async fn run(request: RunProcessRequest) -> RequestResult {
    // TODO(sean) we can infer shell as above for execute if no executable is provided.
    let mut cmd = Command::new(&request.executable);
    #[cfg(target_os = "windows")]
    cmd.creation_flags(windows::Win32::System::Threading::DETACHED_PROCESS.0);

    if let Some(working_directory) = request.working_directory {
        cmd.current_dir(working_directory);
    } else if let Ok(working_directory) = std::env::current_dir() {
        cmd.current_dir(working_directory);
    }
    for arg in request.arguments {
        cmd.arg(arg);
    }
    for var in request.env {
        cmd.env(var.key.clone(), var.value());
    }

    let output = cmd
        .output()
        .await
        .map_err(|_| anyhow!("Failed running command: {:?}", request.executable))?;

    RequestResult::Ok(Box::new(ServerOriginatedSubMessage::RunProcessResponse(
        RunProcessResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: String::from_utf8_lossy(&output.stderr).to_string(),
            exit_code: output.status.code().unwrap_or(0),
        },
    )))
}

pub async fn write() -> RequestResult {
    RequestResult::error("PseudoterminalWriteRequest is deprecated".to_string())
}
