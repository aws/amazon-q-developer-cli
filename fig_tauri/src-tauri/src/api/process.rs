use tokio::process::Command;

use fig_proto::fig::server_originated_message::Submessage as ServerOriginatedSubMessage;
use fig_proto::fig::{
    PseudoterminalExecuteRequest, PseudoterminalExecuteResponse, PseudoterminalWriteRequest,
    RunProcessRequest, RunProcessResponse,
};

use super::{ResponseKind, ResponseResult};
use crate::response_error;

// TODO(mia): implement actual pseudoterminal stuff
pub async fn execute(request: PseudoterminalExecuteRequest, _: i64) -> ResponseResult {
    let mut cmd = Command::new("/bin/bash");
    cmd.arg("--noprofile")
        .arg("--norc")
        .arg("-c")
        .arg(request.command);
    if let Some(working_directory) = request.working_directory {
        cmd.current_dir(working_directory);
    } else if let Ok(working_directory) = std::env::current_dir() {
        cmd.current_dir(working_directory);
    }
    for var in request.env {
        cmd.env(var.key.clone(), var.value());
    }
    let output = cmd
        .output()
        .await
        .map_err(response_error!("Failed running command"))?;

    Ok(ResponseKind::Message(Box::new(
        ServerOriginatedSubMessage::PseudoterminalExecuteResponse(PseudoterminalExecuteResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: if output.stderr.is_empty() {
                None
            } else {
                Some(String::from_utf8_lossy(&output.stderr).to_string())
            },
            exit_code: output.status.code(),
        }),
    )))
}

pub async fn run(request: RunProcessRequest, _: i64) -> ResponseResult {
    let mut cmd = Command::new(request.executable);
    if let Some(working_directory) = request.working_directory {
        cmd.current_dir(working_directory);
    } else if let Ok(working_directory) = std::env::current_dir() {
        cmd.current_dir(working_directory);
    }
    for arg in request.arguments {
        cmd.arg(arg);
    }
    for var in request.env {
        cmd.env(var.key.clone(), var.value());
    }

    let output = cmd
        .output()
        .await
        .map_err(response_error!("Failed running command"))?;

    Ok(ResponseKind::Message(Box::new(
        ServerOriginatedSubMessage::RunProcessResponse(RunProcessResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: String::from_utf8_lossy(&output.stderr).to_string(),
            exit_code: output.status.code().unwrap_or(0),
        }),
    )))
}

pub async fn write(_: PseudoterminalWriteRequest, _: i64) -> ResponseResult {
    Err(ResponseKind::Error(
        "PseudoterminalWriteRequest is deprecated".to_string(),
    ))
}
