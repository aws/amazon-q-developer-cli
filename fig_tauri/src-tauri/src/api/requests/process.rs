use std::process::Stdio;

use anyhow::anyhow;
use fig_proto::fig::server_originated_message::Submessage as ServerOriginatedSubMessage;
use fig_proto::fig::{
    PseudoterminalExecuteRequest,
    PseudoterminalExecuteResponse,
    RunProcessRequest,
    RunProcessResponse,
};
use tokio::process::Command;

use super::{
    RequestResult,
    RequestResultImpl,
};
use crate::native::{
    SHELL,
    SHELL_ARGS,
};

// TODO(mia): implement actual pseudoterminal stuff
pub async fn execute(request: PseudoterminalExecuteRequest) -> RequestResult {
    let mut cmd = Command::new(SHELL);
    cmd.args(SHELL_ARGS);
    cmd.stdin(Stdio::inherit());

    cfg_if::cfg_if!(
        if #[cfg(target_os="windows")] {
            // account for weird behavior passing in commands containing && to WSL
            cmd.args(request.command.split(' ').collect::<Vec<&str>>());
        } else {
            cmd.arg(&request.command);
        }
    );

    if let Some(working_directory) = request.working_directory {
        cmd.current_dir(working_directory);
    }

    for var in request.env {
        cmd.env(var.key.clone(), var.value());
    }

    let output = cmd
        .output()
        .await
        .map_err(|_| anyhow!("Failed running command: {:?}", request.command))?;

    RequestResult::Ok(Box::new(ServerOriginatedSubMessage::PseudoterminalExecuteResponse(
        PseudoterminalExecuteResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: if output.stderr.is_empty() {
                None
            } else {
                Some(String::from_utf8_lossy(&output.stderr).to_string())
            },
            exit_code: output.status.code(),
        },
    )))
}

pub async fn run(request: RunProcessRequest) -> RequestResult {
    let mut cmd = Command::new(&request.executable);
    if let Some(working_directory) = request.working_directory {
        cmd.current_dir(working_directory);
    } else if let Ok(working_directory) = std::env::current_dir() {
        cmd.current_dir(working_directory);
    }
    for arg in request.arguments {
        cmd.arg(arg);
    }
    for var in request.env {
        cmd.env(var.key.clone(), var.value());
    }

    let output = cmd
        .output()
        .await
        .map_err(|_| anyhow!("Failed running command: {:?}", request.executable))?;

    RequestResult::Ok(Box::new(ServerOriginatedSubMessage::RunProcessResponse(
        RunProcessResponse {
            stdout: String::from_utf8_lossy(&output.stdout).to_string(),
            stderr: String::from_utf8_lossy(&output.stderr).to_string(),
            exit_code: output.status.code().unwrap_or(0),
        },
    )))
}

pub async fn write() -> RequestResult {
    RequestResult::error("PseudoterminalWriteRequest is deprecated".to_string())
}
