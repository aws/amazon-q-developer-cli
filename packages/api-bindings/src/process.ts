import { sendRunProcessRequest } from "./requests.js";

export async function run({
  executable,
  args,
  environment,
  workingDirectory,
  terminalSessionId,
  timeout,
}: {
  executable: string;
  args: string[];
  environment?: Record<string, string | undefined>;
  workingDirectory?: string;
  terminalSessionId?: string;
  timeout?: number;
}) {
  const env = environment ?? {};
  return sendRunProcessRequest({
    executable,
    arguments: args,
    env: Object.keys(env).map((key) => ({ key, value: env[key] })),
    workingDirectory,
    terminalSessionId,
    timeout: timeout
      ? {
          nanos: Math.floor((timeout % 1000) * 1_000_000_000),
          secs: Math.floor(timeout / 1000),
        }
      : undefined,
  });
}
