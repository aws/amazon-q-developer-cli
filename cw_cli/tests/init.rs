#[cfg(not(windows))]
use std::io::Write;
#[cfg(not(windows))]
use std::process::{
    Command,
    Stdio,
};

#[cfg(not(windows))]
use assert_cmd::prelude::*;
use eyre::Context;
use paste::paste;

macro_rules! init_test {
    ($shell:expr, $stage:expr, $file:expr, [$exe:expr, $($arg:expr),*]) => {
        paste! {
            #[cfg(not(windows))]
            fn [<init_output_ $shell _ $stage _ $file>]() -> Result<String, Box<dyn std::error::Error>> {
                let mut cmd = Command::cargo_bin("cw_cli")?;
                cmd.arg("init").arg($shell).arg($stage).arg("--rcfile").arg($file);
                cmd.env("CW_INIT_SNAPSHOT_TEST", "1");
                let out = cmd.assert().success().get_output().stdout.clone();
                Ok(String::from_utf8(out)?)
            }

            #[test]
            #[cfg(not(windows))]
            fn [<init_snapshot_ $shell _ $stage _ $file>]() -> Result<(), Box<dyn std::error::Error>> {
                let init = [<init_output_ $shell _ $stage _ $file>]()?;

                insta::assert_snapshot!(init);

                Ok(())
            }

            #[test]
            #[cfg(not(windows))]
            fn [<init_lint_ $shell _ $stage _ $file>]() -> Result<(), Box<dyn std::error::Error>> {
                // Ignore fish post since idk it doesn't work on CI
                if $exe == "fish" && $stage == "post" {
                    return Ok(());
                }

                // Ignore all fish in brazil
                if $exe == "fish" && std::env::var_os("BRAZIL_BUILD_HOME").is_some() {
                    return Ok(());
                }

                if std::env::var_os("CW_BUILD_SKIP_SHELL_TESTS").is_some() {
                    return Ok(());
                }

                let init = [<init_output_ $shell _ $stage _ $file>]()?;

                let mut cmd = Command::new($exe);
                cmd$(.arg($arg))*.stdin(Stdio::piped()).stdout(Stdio::piped()).stderr(Stdio::piped());
                cmd.env("CW_INIT_SNAPSHOT_TEST", "1");

                let child = cmd.spawn().context(format!("{} is not installed", $exe))?;
                write!(child.stdin.as_ref().unwrap(), "{}", init)?;
                let output = child.wait_with_output()?;
                if !output.status.success() {
                    let stdout = String::from_utf8(output.stdout)?;
                    let stderr = String::from_utf8(output.stderr)?;
                    println!("stdout: {stdout}");
                    println!("stderr: {stderr}");

                    // Write shell version to stdout
                    let mut cmd = Command::new($exe);
                    cmd.arg("--version");
                    let out = cmd.output()?;
                    println!("Linter {} version: {}", $exe, String::from_utf8(out.stdout)?);

                    panic!(
                        "linter returned {}. please run `cargo run -p cw_cli -- init {} {} --rcfile {} | {} {}`",
                        output.status, $shell, $stage, $file, $exe, [$($arg),*].join(" ")
                    );
                }

                Ok(())
            }
        }
    };
}

// bash
init_test!("bash", "pre", "bashrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("bash", "pre", "bash_profile", ["shellcheck", "-s", "bash", "-"]);
init_test!("bash", "post", "bashrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("bash", "post", "bash_profile", ["shellcheck", "-s", "bash", "-"]);

// zsh
init_test!("zsh", "pre", "zshrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("zsh", "pre", "zprofile", ["shellcheck", "-s", "bash", "-"]);
init_test!("zsh", "post", "zshrc", ["shellcheck", "-s", "bash", "-"]);
init_test!("zsh", "post", "zprofile", ["shellcheck", "-s", "bash", "-"]);

// fish
init_test!("fish", "pre", "00_fig_pre", ["fish", "--no-execute"]);
init_test!("fish", "post", "99_fig_post", ["fish", "--no-execute"]);
