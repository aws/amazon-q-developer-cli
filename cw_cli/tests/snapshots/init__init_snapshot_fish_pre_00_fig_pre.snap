---
source: cw_cli/tests/init.rs
expression: init
---
set -g CW_SHELL "/bin/zsh"
set -g SHOULD_CWTERM_LAUNCH "0"
command mkdir -p ~/.local/bin >/dev/null

builtin contains $HOME/.local/bin $PATH
or set --append PATH $HOME/.local/bin

builtin contains $HOME/.local/bin $PATH
or set --append PATH $HOME/.local/bin

if not test -z "$FIG_NEW_SESSION"
    set --erase CWTERM_SESSION_ID
    set --erase CW_TERM
    set --erase FIG_ENV_VAR
    set --erase FIG_NEW_SESSION
end

# Load parent from env variables
if test -n "$FIG_SET_PARENT"; and test -z "$LC_FIG_SET_PARENT"
    set --export FIG_SET_PARENT $LC_FIG_SET_PARENT
end
if test -n "$FIG_PARENT"; and test -z "$FIG_SET_PARENT"
    set --export FIG_PARENT $FIG_SET_PARENT
end

if test -z "$SHOULD_CWTERM_LAUNCH"
    # 0 = Yes, 1 = No, 2 = Fallback to CW_TERM
    cw _ should-figterm-launch 1>/dev/null 2>&1
    set SHOULD_CWTERM_LAUNCH $status
end

if test -t 1
    and test -z "$PROCESS_LAUNCHED_BY_FIG"
    and command -v cwterm 1>/dev/null 2>/dev/null
    and test "$SHOULD_CWTERM_LAUNCH" -eq 0 -o \( "$SHOULD_CWTERM_LAUNCH" -eq 2 -a \( -z "$CW_TERM" -o \( -z "$CW_TERM_TMUX" -a -n "$TMUX" \) \) \)

    if test -z "$CW_SHELL"
        set CW_SHELL (cw _ get-shell)
    end
    set FIG_IS_LOGIN_SHELL 0
    if status --is-login
        set FIG_IS_LOGIN_SHELL 1
    end

    # Do not launch cwterm in non-interactive shells (like VSCode Tasks)
    if status --is-interactive
        set CW_TERM_NAME (command basename "$CW_SHELL")" (cwterm)"
        if test -x "$HOME/.local/bin/$CW_TERM_NAME"
            set CW_TERM_PATH "$HOME/.local/bin/$CW_TERM_NAME"
        else
            set CW_TERM_PATH (command -v cwterm || echo "$HOME/.local/bin/cwterm")
        end

        # Need to exec bash because we're using 'exec -a <name>'
        # to set argv[0] and fish's exec doesn't have this option
        exec bash -c "FIG_PARENT=$FIG_PARENT CW_SHELL=$CW_SHELL FIG_IS_LOGIN_SHELL=$FIG_IS_LOGIN_SHELL exec -a \"$CW_TERM_NAME\" \"$CW_TERM_PATH\""
    end
    # else
    #     set -g FIG_DID_NOT_EXEC_FIGTERM 1
end


